File: data/categories-data/src/main/java/app/fit/fitndflow/data/categories/repositoryImpl/CategoriesRepositoryImpl.java
package app.fit.fitndflow.data.categories.repositoryImpl;

import android.content.Context;


import com.fit.fitndflow.app.domain.categories.repository.CategoriesRepository;
import com.fit.fitndflow.app.domain.common.models.CategoryModel;

import java.util.List;

import app.fit.fitndflow.data.categories.datasource.remote.CategoryRemoteDataSource;
import app.fit.fitndflow.data.categories.dto.AddCategoryDto;
import app.fit.fitndflow.data.common.database.dao.CategoryDao;
import app.fit.fitndflow.data.common.dto.CategoryDto;
import app.fit.fitndflow.data.categories.dto.ModifyCategoryDto;
import app.fit.fitndflow.data.categories.model.CategoriesApiInterface;
import app.fit.fitndflow.data.common.mapper.StringInLanguagesMapperKt;
import app.fit.fitndflow.data.common.model.ExcepcionApi;
import app.fit.fitndflow.data.common.datasource.CategoriesAndExercisesLocalDataSource;
import app.fit.fitndflow.data.common.datasource.TrainingLocalDataSource;
import app.fit.fitndflow.data.common.dto.StringInLanguagesDto;
import app.fit.fitndflow.data.common.mapper.CategoryModelMapperKt;
import retrofit2.Response;

public class CategoriesRepositoryImpl implements CategoriesRepository {
    private CategoriesAndExercisesLocalDataSource categoriesAndExercisesLocalDataSource;
    private TrainingLocalDataSource trainingLocalDataSource;
    private CategoryRemoteDataSource categoryRemoteDataSource;
    private CategoryDao categoryDao;
    private Context mContext;
    private boolean isLocalMode = true;

    public CategoriesRepositoryImpl(Context context, CategoryRemoteDataSource categoryRemoteDataSource, CategoryDao categoryDao ,CategoriesAndExercisesLocalDataSource categoriesAndExercisesLocalDataSource, TrainingLocalDataSource trainingLocalDataSource) {
        this.mContext = context;
        this.categoryRemoteDataSource = categoryRemoteDataSource;
        this.categoriesAndExercisesLocalDataSource = categoriesAndExercisesLocalDataSource;
        this.trainingLocalDataSource = trainingLocalDataSource;
        this.categoryDao = categoryDao;
    }

    @Override
    public List<CategoryModel> getCategoryList() throws Exception {
        if (categoriesAndExercisesLocalDataSource.getAvailableCategoryListCache() == null) {
            List<CategoryModel> response;
            try {
                if (isLocalMode) {
                    response = categoryDao.getAllCategories();
                } else {
                    response = categoryRemoteDataSource.getCategoryList();
                }
                categoriesAndExercisesLocalDataSource.replaceAllDataFromCategoryListCache(response);
            } catch (Exception e) {
                e.printStackTrace();
                throw new Exception(e);
            }
        }
        return categoriesAndExercisesLocalDataSource.getAvailableCategoryListCache();
    }

    @Override
    public List<CategoryModel> addNewCategory(String categoryName, String language) throws Exception {
        StringInLanguagesDto stringInLanguages = StringInLanguagesMapperKt.convertToStringInLanguages(language, categoryName);
        List<CategoryDto> response;
        try {
            response = categoryRemoteDataSource.addNewCategory(stringInLanguages);
            categoriesAndExercisesLocalDataSource.replaceAllDataFromCategoryListCache(CategoryModelMapperKt.toModel(response));
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return categoriesAndExercisesLocalDataSource.getAvailableCategoryListCache();
    }

    @Override
    public List<CategoryModel> modifyCategory(String categoryName, String language, int categoryId, String imageUrl) throws Exception {
        StringInLanguagesDto stringInLanguages = StringInLanguagesMapperKt.convertToStringInLanguages(language, categoryName);
        List<CategoryDto> response;
        try {
            response = categoryRemoteDataSource.modifyCategory(stringInLanguages, categoryId, imageUrl);
            categoriesAndExercisesLocalDataSource.replaceAllDataFromCategoryListCache(CategoryModelMapperKt.toModel(response));
            trainingLocalDataSource.cleanCache();
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return categoriesAndExercisesLocalDataSource.getAvailableCategoryListCache();
    }

    @Override
    public List<CategoryModel> deleteCategory(Integer categoryId) throws Exception {
        List<CategoryDto> response;
        try {
            response = categoryRemoteDataSource.deleteCategory(categoryId);
            categoriesAndExercisesLocalDataSource.replaceAllDataFromCategoryListCache(CategoryModelMapperKt.toModel(response));
            trainingLocalDataSource.cleanCache();
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e);
        }
        return categoriesAndExercisesLocalDataSource.getAvailableCategoryListCache();
    }
}

File: data/categories-data/src/main/java/app/fit/fitndflow/data/categories/repositoryImpl/CategoriesRepositoryImpl.kt
package app.fit.fitndflow.data.categories.repositoryImpl

import app.fit.fitndflow.data.categories.datasource.local.InitialExercisesCreatorHelper
import app.fit.fitndflow.data.categories.datasource.remote.CategoryRemoteDataSource
import app.fit.fitndflow.data.common.database.DataBase
import app.fit.fitndflow.data.common.database.dao.CategoryDao
import app.fit.fitndflow.data.common.database.dao.ExerciseDao
import app.fit.fitndflow.data.common.database.entities.CategoryEntity
import app.fit.fitndflow.data.common.database.entities.ExerciseEntity
import app.fit.fitndflow.data.common.database.mapper.toModel
import app.fit.fitndflow.data.common.datasource.local.CategoriesAndExercisesCacheLocalDataSource
import app.fit.fitndflow.data.common.datasource.local.SharedPrefsLocalDataSource
import app.fit.fitndflow.data.common.datasource.local.TrainingCacheLocalDataSource
import app.fit.fitndflow.data.common.mapper.CategoryModelMapperKt.Companion.toModel
import app.fit.fitndflow.data.common.mapper.convertToStringInLanguages
import com.fit.fitndflow.app.domain.categories.repository.CategoriesRepository
import com.fit.fitndflow.app.domain.common.models.CategoryModel

class CategoriesRepositoryImpl(
    private val categoryRemoteDataSource: CategoryRemoteDataSource,
    private val categoryDao: CategoryDao,
    private val exerciseDao: ExerciseDao,
    private val categoriesAndExercisesCacheLocalDataSource: CategoriesAndExercisesCacheLocalDataSource,
    private val trainingCacheLocalDataSource: TrainingCacheLocalDataSource,
    private val initialExercisesCreatorHelper: InitialExercisesCreatorHelper,
    private val sharedPrefsLocalDataSource: SharedPrefsLocalDataSource,
) : CategoriesRepository {
    private val isLocalMode
        get() = sharedPrefsLocalDataSource.getIsLocal()

    override suspend fun getCategoryList(): List<CategoryModel> {
        if (categoriesAndExercisesCacheLocalDataSource.getAvailableCategoryListCache() == null) {
            val response: List<CategoryModel>
            try {
                if (isLocalMode) {
                    response = categoryDao.getAllCategories().map { it.toModel() }
                } else {
                    response = categoryRemoteDataSource.getCategoryList()
                }
                categoriesAndExercisesCacheLocalDataSource.replaceAllDataFromCategoryListCache(
                    response
                )
            } catch (e: Exception) {
                e.printStackTrace()
                throw Exception(e)
            }
        }
        return categoriesAndExercisesCacheLocalDataSource.getAvailableCategoryListCache().orEmpty()
    }

    override suspend fun addNewCategory(
        categoryName: String,
        language: String
    ): List<CategoryModel> {
        val stringInLanguages = convertToStringInLanguages(language, categoryName)
        val response: List<CategoryModel>
        try {
            if (isLocalMode) {
                categoryDao.insertAllCategories(
                    listOf(
                        CategoryEntity(
                            nameEs = stringInLanguages.spanish.orEmpty(),
                            nameEn = stringInLanguages.english.orEmpty()
                        )
                    )
                )
                response = categoryDao.getAllCategories().map { it.toModel() }
            } else {
                response = toModel(categoryRemoteDataSource.addNewCategory(stringInLanguages))
            }
            categoriesAndExercisesCacheLocalDataSource.replaceAllDataFromCategoryListCache(response)
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return categoriesAndExercisesCacheLocalDataSource.getAvailableCategoryListCache().orEmpty()
    }

    override suspend fun modifyCategory(
        categoryName: String,
        language: String,
        categoryId: Int,
        imageUrl: String?
    ): List<CategoryModel> {
        val stringInLanguages = convertToStringInLanguages(language, categoryName)
        val response: List<CategoryModel>
        try {
            if (isLocalMode) {
                categoryDao.updateCategory(
                    categoryId = categoryId,
                    nameEs = stringInLanguages.spanish.orEmpty(),
                    nameEn = stringInLanguages.english.orEmpty(),
                )
                response = categoryDao.getAllCategories().map { it.toModel() }
            } else {
                response = toModel(
                    categoryRemoteDataSource.modifyCategory(
                        stringInLanguages,
                        categoryId,
                        imageUrl
                    )
                )
            }
            categoriesAndExercisesCacheLocalDataSource.replaceAllDataFromCategoryListCache(response)
            trainingCacheLocalDataSource.cleanCache()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return categoriesAndExercisesCacheLocalDataSource.getAvailableCategoryListCache().orEmpty()
    }

    override suspend fun deleteCategory(categoryId: Int): List<CategoryModel> {
        val response: List<CategoryModel>
        try {
            if (isLocalMode) {
                categoryDao.deleteCategory(categoryId)
                response = categoryDao.getAllCategories().map { it.toModel() }
            } else {
                response = toModel(categoryRemoteDataSource.deleteCategory(categoryId))
            }
            categoriesAndExercisesCacheLocalDataSource.replaceAllDataFromCategoryListCache(response)
            trainingCacheLocalDataSource.cleanCache()
        } catch (e: Exception) {
            e.printStackTrace()
            throw Exception(e)
        }
        return categoriesAndExercisesCacheLocalDataSource.getAvailableCategoryListCache().orEmpty()
    }

    override suspend fun migrateToLocal() {
        val remoteCategories = categoryRemoteDataSource.getCategoryList()

        val categoryEntities = remoteCategories.mapNotNull { category ->
            category.id?.let { id ->
                CategoryEntity(
                    id = id,
                    nameEs = category.name.spanish,
                    nameEn = category.name.english
                )
            }
        }

        val exerciseEntities = remoteCategories.flatMap { category ->
            val categoryId = category.id ?: return@flatMap emptyList()
            category.exerciseList.orEmpty().mapNotNull { exercise ->
                exercise.id?.let { exerciseId ->
                    ExerciseEntity(
                        id = exerciseId,
                        categoryId = categoryId,
                        nameEs = exercise.name.spanish,
                        nameEn = exercise.name.english,
                        recordHeight = exercise.record?.kg ?: 0.0,
                        recordReps = exercise.record?.reps ?: 0,
                        firstReps = exercise.lastFirstSerie?.reps ?: 0,
                        firstWeight = exercise.lastFirstSerie?.kg ?: 0.0
                    )
                }
            }
        }
        categoryDao.insertAllCategories(categoryEntities)
        exerciseDao.insertExercises(exerciseEntities)

        categoriesAndExercisesCacheLocalDataSource.cleanCache()
        trainingCacheLocalDataSource.cleanCache()
        sharedPrefsLocalDataSource.saveIsLocal()
        sharedPrefsLocalDataSource.saveInitialDataCreated()
        sharedPrefsLocalDataSource.saveApiKey(null)
    }

    override suspend fun createInitialData() {
        sharedPrefsLocalDataSource.saveIsLocal()
        initialExercisesCreatorHelper.insertInitialCategories()
        sharedPrefsLocalDataSource.saveInitialDataCreated()
        return
    }

    override fun isInitialDataCreated() = sharedPrefsLocalDataSource.getIsInitialDataCreated()
}
